Add gost tls13

VladGud opened the pull request:
### Add GOST TLS 1.3 (RFC 9367) support to OpenSSL

This patch introduces support for Russian cryptographic algorithms in TLS 1.3 as specified in [RFC 9367](https://www.rfc-editor.org/rfc/rfc9367). This work addresses the request tracked in [openssl/openssl#23758](https://github.com/openssl/openssl/issues/23758).

#### Changes
- **New TLS 1.3 cipher suites:**
  - `TLS_GOSTR341112_256_WITH_KUZNYECHIK_MGM_L`
  - `TLS_GOSTR341112_256_WITH_MAGMA_MGM_L`
  - `TLS_GOSTR341112_256_WITH_KUZNYECHIK_MGM_S`
  - `TLS_GOSTR341112_256_WITH_MAGMA_MGM_S`

- **New ciphers and OIDs:**
  - `magma-mgm`
  - `kuznyechik-mgm`

- **EVP updates:**
  - Added parameters `OSSL_CIPHER_PARAM_TLSTREE` and `OSSL_CIPHER_PARAM_TLSTREE_MODE`.

- **TLSTREE key diversification:**  
  - New flags: `TLS1_TLSTREE`, `TLS1_TLSTREE_S`, `TLS1_TLSTREE_L`.  
  - Integrated into `tls13_meth.c` and `rec_layer_s3.c` with calls to `EVP_CIPHER_CTX_set_params` to pass sequence numbers and configure TLSTREE mode.
  - Previously TLSTREE was only used for TLS 1.2 GOST cipher suites. This patch extends support to TLS 1.3, including distinction between "light" and "strong" modes (L/S flags). Required non-trivial changes in `tls13_meth.c` and `rec_layer_s3.c`.

- **SSL/Cipher layer extensions:**
  - New `SSL_ENC_*` indices for MGM-based ciphers.  
  - Updated `ssl_ciph.c` for proper cipher descriptions and filtering.  

- **AEAD tag lengths for GOST TLS 1.3:**
  - `EVP_MAGMA_TLS_TAG_LEN = 8`
  - `EVP_KUZNYECHIK_TLS_TAG_LEN = 16`
  - Integrated into `tls13_enc.c`.

- **Signature handling updates:**
* Signature handling was updated to reuse OpenSSL’s existing hardcoded GOST checks in the gost-provider.
* This was achieved by defining algorithms under the same names as in the gost-engine.
* All calls to EVP_PKEY_get_id were replaced with EVP_PKEY_is_a to ensure compatibility with both providers and engines.
The sigalg_allowed logic was refined to: exclude GOST2001 in TLS 1.3 and allow GOST2012-256/GOST2012-512 as valid signature algorithms.

#### Tests
All TLS 1.3 GOST functionality introduced in this patch was tested in a separate [PR1](https://github.com/gost-engine/engine/pull/482) in the gost-engine repository, which includes integration tests and interoperability checks. Additionally, a follow-up [PR2](https://github.com/gost-engine/engine/pull/489) extends the test coverage and provides several minor fixes.

The final `gost-engine` testing against the current changes from this PR is currently available in this fork commit: https://github.com/VladGud/engine/commit/469a35f900787f89c3a2d8cc6ba514b90ae12d31. These changes will be transferred to the upstream `gost-engine` repository once the changes in this PR are approved. The fork will continue to be updated in sync with the changes in this PR, which allows us to test and update `gost-engine` more quickly and independently while the review is ongoing.

VladGud commented on test/recipes/95-test_external_gost_engine_data/gost_engine.sh:
```diff
@@ -51,4 +51,5 @@ export CTEST_OUTPUT_ON_FAILURE=1
 make test
-make tcl_tests
+make tcl_tests_engine
+make tcl_tests_provider
```
Yes, we are aware. In this case, `tcl_tests_engine` is just a rename of the `tcl_tests` target. Before this PR, the same tests ran in OpenSSL and passed. They also covered the engine tests. If needed, we can remove it, but it may be better to do this step by step.

In general, gost-engine also implements TLS 1.3 through the OpenSSL provider interface.

VladGud replied:
Okay. Then we will update gost-engine as well and drop the removed interfaces in the same way. Thank you!

xnox commented:
also see https://github.com/openssl/openssl/pull/29925

* Can all of these additions be wrapped with ifndef OPENSSL_NO_GOST?

* Are any of the previous workarounds, bugs, legacy, compat options / extensions / curves / algorithms can be dropped? It seems like many of them are buggy and no longer in use.

* Note many codepaths are live even in OPENSSL_NO_GOST builds, and need fixing.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

Commit:
Add support for GOST TLS 1.3 implementation
This patch introduces support for Russian cryptographic algorithms in TLS 1.3 as specified in [RFC 9367](https://www.rfc-editor.org/rfc/rfc9367).

Changes:
- **New TLS 1.3 cipher suites:**
  - `TLS_GOSTR341112_256_WITH_KUZNYECHIK_MGM_L`
  - `TLS_GOSTR341112_256_WITH_MAGMA_MGM_L`
  - `TLS_GOSTR341112_256_WITH_KUZNYECHIK_MGM_S`
  - `TLS_GOSTR341112_256_WITH_MAGMA_MGM_S`

- **New ciphers and OIDs:**
  - `magma-mgm`
  - `kuznyechik-mgm`

- **EVP updates:**
  - Added parameters `OSSL_CIPHER_PARAM_TLSTREE` and `OSSL_CIPHER_PARAM_TLSTREE_MODE`.

- **TLSTREE key diversification:**
  - New flags: `TLS1_TLSTREE`, `TLS1_TLSTREE_S`, `TLS1_TLSTREE_L`.
  - Integrated into `tls13_meth.c` and `rec_layer_s3.c` with calls to `EVP_CIPHER_CTX_set_params` to pass sequence numbers and configure TLSTREE mode.
  - Previously TLSTREE was only used for TLS 1.2 GOST cipher suites. This patch extends support to TLS 1.3, including distinction between "light" and "strong" modes (L/S flags). Required non-trivial changes in `tls13_meth.c` and `rec_layer_s3.c`.

- **SSL/Cipher layer extensions:**
  - New `SSL_ENC_*` indices for MGM-based ciphers.
  - Updated `ssl_ciph.c` for proper cipher descriptions and filtering.

- **AEAD tag lengths for GOST TLS 1.3:**
  - `EVP_MAGMA_TLS_TAG_LEN = 8`
  - `EVP_KUZNYECHIK_TLS_TAG_LEN = 16`
  - Integrated into `tls13_enc.c`.

- **Signature handling updates:**
* Signature handling was updated to reuse OpenSSL’s existing hardcoded GOST checks in the gost-provider.
* This was achieved by defining algorithms under the same names as in the gost-engine.
* All calls to EVP_PKEY_get_id were replaced with EVP_PKEY_is_a to ensure compatibility with both providers and engines.
The sigalg_allowed logic was refined to: exclude GOST2001 in TLS 1.3 and allow GOST2012-256/GOST2012-512 as valid signature algorithms.

VladGud commented:
@slontis @esyr @t8m

Apologies for the delay. Could you please take another look at this PR?

The required changes have now been made in the `gost-engine` project. The `gost-provider` implementation is complete and is no longer tied to the legacy `ENGINE` logic: https://github.com/gost-engine/engine/commit/a34a05e29a1b402335a8faaafba75dbdd192aadd.

The tests can be checked using this small fork of the `gost-engine` project: https://github.com/VladGud/engine/commit/469a35f900787f89c3a2d8cc6ba514b90ae12d31. I'm keeping this fork in sync with the current changes in this PR, so we can test and update `gost-engine` quickly while the review is ongoing.

Once this PR is approved/merged, we'll update the upstream `gost-engine` repository with the final changes. After that, we'll send a separate PR to OpenSSL with full GOST TLS 1.3 tests covering the functionality introduced here.

mattcaswell commented on ssl/record/methods/tls13_meth.c:
```diff
@@ -80,6 +81,31 @@ static int tls13_set_crypto_state(OSSL_RECORD_LAYER *rl, int level,
     }
+
+#ifndef OPENSSL_NO_GOST
+    if (!rl->isdtls && rl->tlstree) {
```
```suggestion
    if (!rl->isdtls && rl->tlstree != 0) {
```

mattcaswell commented on ssl/record/methods/tls13_meth.c:
```diff
@@ -164,6 +190,20 @@ static int tls13_cipher(OSSL_RECORD_LAYER *rl, TLS_RL_RECORD *recs,
         nonce[offset + loop] = staticiv[offset + loop] ^ seq[loop];
 
+#ifndef OPENSSL_NO_GOST
+    if (!rl->isdtls && rl->tlstree) {
```
```suggestion
    if (!rl->isdtls && rl->tlstree != 0) {
```

mattcaswell commented on ssl/record/rec_layer_s3.c:
```diff
@@ -1348,7 +1357,7 @@ int ssl_set_new_record_layer(SSL_CONNECTION *s, int version,
             &stream_mac);
 
     if (tlstree)
```
This is pre-existing but really should be:

```suggestion
    if (tlstree != 0)
```

mattcaswell commented on ssl/record/rec_layer_s3.c:
```diff
@@ -1339,6 +1340,14 @@ int ssl_set_new_record_layer(SSL_CONNECTION *s, int version,
+    else if (s->session != NULL && s->session->cipher != NULL)
+        algorithm2 = s->session->cipher->algorithm2;
+    else if (s->psksession != NULL && s->psksession->cipher != NULL)
+        algorithm2 = s->psksession->cipher->algorithm2;
```
I assume all of this is related to early-data. Please can you add some kind of comment about this?

VladGud replied:
Yes. Fixed: added a comment.

mattcaswell commented on ssl/statem/statem_lib.c:
```diff
@@ -1948,6 +1945,15 @@ static int is_tls13_capable(const SSL_CONNECTION *s)
+
+        /*
+         * TODO: There must be an opportunity to check that the sig algs are
+         * consistent with the keys in case if the sig algs are provider-based.
```
I don't understand what you mean by this. Please can you explain further.

VladGud replied:
This TODO refers to a provider-aware key/sigalg consistency check. It is analogous in purpose to the EC-specific tls_check_sigalg_curve() check, but not necessarily identical in mechanism.

For built-in EC keys, TLS 1.3 can verify that the configured signature algorithms are consistent with the key’s curve. For provider-based sigalgs, there is no equivalent check today.

The concrete gaps are:
- In ssl_setup_sigalgs(), provider-based sigalgs are stored with curve == NID_undef.
- As a result, the SSL layer does not retain curve / parameter-set information for provider sigalgs.
- There is no mapping between a provider-backed key instance (keymgmt alg + paramset) and the TLS sigalg code points registered through the TLS sigalgs capability.
- Because of that, SSL can see that a provider exposes some TLS signature algorithms in general, but it cannot correctly verify that a specific key matches a specific TLS 1.3 signature algorithm.

This is important for cases like GOST TLS 1.3, where the same key algorithm is associated with multiple TLS code points corresponding to different paramset / curves:

https://github.com/VladGud/engine/blob/a34a05e29a1b402335a8faaafba75dbdd192aadd/gost_prov_tls.c#L78

So the TODO is about adding a way to associate a provider-backed key with the TLS sigalg code points it actually supports, in the same spirit as tls_check_sigalg_curve() for built-in EC
keys.

vdukhovni replied:
The correct way to ensure accurate signature algorithm matching in `tls12_check_peer_sigalg()` is to NOT bundle multiple signature algorithms under a single signature algorithm differentiated only the by the "curve" name.

While that's the case for the existing "ec" algorithms, it is not a good idea going forward.  Instead, each of a **small** number of supported variants in a family gets its own algorithm name (e.g. ML-DSA-44/65/87), the "curve" issue DOES NOT arise.

If you're tempted to mimmic the existing practice in the "ec" family, with its panoply of curves (most now deprecated). DON'T!

VladGud replied:
This could indeed be a good approach.

However, splitting this into multiple signature algorithms would also require splitting the keymgmt side accordingly (i.e. introducing additional key algorithm names).

My concern is that OpenSSL already has existing SN/LN names for the GOST signature algorithms, and there is already a fair amount of OpenSSL logic built around those names. For example, there is existing GOST-specific handling such as the buffer reversal logic:
```c
#ifndef OPENSSL_NO_GOST
   {
       if (EVP_PKEY_is_a(pkey, SN_id_GostR3410_2001)
           || EVP_PKEY_is_a(pkey, SN_id_GostR3410_2012_256)
           || EVP_PKEY_is_a(pkey, SN_id_GostR3410_2012_512)) {
           if ((gost_data = OPENSSL_malloc(len)) == NULL)
               goto err;
           BUF_reverse(gost_data, data, len);
           data = gost_data;
       }
   }
#endif
```

This logic is already reused in the current PR. It will also be reused for GOST TLS 1.2 in the other PR: https://github.com/openssl/openssl/pull/29921. Because of that, introducing new names for the new keymgmt / sigalg combinations in OpenSSL, and then adjusting this existing logic, looks more expensive to me.

I also have some questions about OSSL_PKEY_PARAM_GROUP_NAME and OSSL_CAPABILITY_TLS_GROUP_ALG. On the KEX/KEM side, the TLS group can specify the keymgmt algorithm via OSSL_CAPABILITY_TLS_GROUP_ALG, and libssl then passes the group name through OSSL_PKEY_PARAM_GROUP_NAME when generating keys or parameters. This effectively allows one GOST key algorithm to work with multiple curves without introducing separate key algorithm names for each variant, so perhaps it would be useful to have something similar for TLS signature algorithms as well.

That said, I may still be missing something or misunderstanding your point, so apologies if that is the case.

vdukhovni replied:
See existing example in `providers/implementations/keymgmt/ml_dsa_kmgmt.c`, one implementation, but indeed multiple (3) keymgmt handles, generated via a macro.

```
MAKE_KEYMGMT_FUNCTIONS(44);                                                                                             
MAKE_KEYMGMT_FUNCTIONS(65);                                                                                             
MAKE_KEYMGMT_FUNCTIONS(87);                                                                                             
```

New provided TLS algorithms are TLS 1.3 only, and the legacy GOST code for TLS 1.2 should ideally simply be deleted.  OpenSSL TLS stack should have no special knowledge of GOST, just like there's no specific knowledge of ML-DSA.

All data about GOST is obtained from the provider, the implementation treats each of the GOST signature algorithms just like it would any other new provider-based signature algorithm, and the same for key exchange groups (see the ML-KEM provider).

vdukhovni replied:
In TLS 1.3 the `supported_groups` extension is scoped to only the key exchange groups, while supported signature algorithms are negotiated directly, with no relation to supported curves.

This is apparent from RFC8446 Section 4.2.6, e.g.

```
   -  ECDSA signature schemes align with TLS 1.2's ECDSA hash/signature
      pairs.  However, the old semantics did not constrain the signing
      curve.  If TLS 1.2 is negotiated, implementations MUST be prepared
      to accept a signature that uses any curve that they advertised in
      the "supported_groups" extension.
```
That comments on the difference with TLS 1.2, where supported_curves applied to signatures, but that is now obsolete.

VladGud replied:
If we remove the legacy GOST TLS 1.2 code, how should GOST TLS 1.2 be supported in the current upstream codebase, given that the provider-based implementation currently only covers TLS 1.3?

There are also still open questions about moving the code being removed, for example buffer byte-order conversion, into the provider side. In that case, we would have to support different signature algorithms, because TLS signatures and non-TLS signatures use different representations. Those algorithms would also need to have different key NIDs. This adds extra overhead that we would very much prefer to avoid.

Regarding TLS 1.3 signature algorithms, we understand the general approach. However, the concern is about the practical implementation of provider-side signatures. The documentation does not state anywhere that provider signatures must be based on TLS 1.3 signature algorithms. From the implementation, my expectation was that mapping provider signatures through TLS signature capabilities would define the required TLS 1.3 representation.

However, it seems that TLS signature capabilities currently do not provide enough information about the actual signature algorithm. One possible improvement would be to add something like OSSL_CAPABILITY_TLS_SIGALG_GROUP_NAME. With this capability, it would be possible to determine which curve actually corresponds to a given signature_algorithms entry. Then, in ssl_setup_sigalgs, provider-based algorithms would not need to set curve = NID_undef.

vdukhovni replied:
TLS 1.2 is frozen, I see no reason to add support for updated versions of GOST to TLS 1.2.
So my take is that new GOST algorithms in TLS 1.2 are NOT supported, at least not by the "upstream" OpenSSL.
These only go into a provider that vends TLS 1.3 algorithms, that don't rely on additional "group" properties being available.

vdukhovni replied:
In any case it makes sense to set aside the TLS 1.2 question, and focus on the TLS 1.3 implementation. If you want to make a case for TLS 1.2, that can be done in a separate later design discussion.

Mironenko replied:
@vdukhovni, I would just like to clarify the expected direction here.

GOST TLS 1.2 cipher suites are standardized in [RFC9189](https://datatracker.ietf.org/doc/html/rfc9189) and were supported in OpenSSL 3.x via gost-engine using the legacy ENGINE interface. Since ENGINE support is being removed in OpenSSL 4.x, am I understanding correctly that upstream OpenSSL does not intend to support RFC9189-based GOST TLS 1.2 via the provider interface either?

In other words, is the expectation that provider-based GOST support in OpenSSL 4.x should be limited strictly to TLS 1.3?

vdukhovni replied:
I am not making an official statement for the project, just a personal viewpoint.  With that disclaimer out of the way, yes, my preference is that provider-based GOST be TLS 1.3 only, and then the legacy code for TLS 1.2 can be removed when reasonably convenient.

VladGud replied:
@mattcaswell, could you please help clarify what behavior is expected here?

From the recent discussion, we understood that provider signature implementations are expected to map more or less 1:1 to TLS 1.3 signature algorithms. However, this is not entirely obvious to us from the current provider/TLS capability model, so we would like to make sure we are interpreting the intended design correctly.

For GOST, this seems somewhat difficult in practice. If each code point needs to be represented as a separate signature algorithm, and also as a separate keymgmt identity, that would make the implementation noticeably more complex.

The reason is that some existing OpenSSL logic is already tied to the current GOST algorithm identities/names, and we are currently reusing that logic. At the same time, this does not seem easy to move cleanly into the provider side, so splitting everything into separate per-code-point signature algorithms would make such reuse significantly harder.

Because of that, it would be very helpful to understand which model is expected here:
1) Should provider signature implementations be modeled directly as TLS 1.3 signature algorithms?
2) Or is it acceptable for one provider signature/key type to correspond to multiple TLS signature algorithm code points, with some additional way to disambiguate the concrete variant?

One possible direction we had in mind for the second model would be to extend the TLS sigalg capability with some extra parameter describing the concrete variant, for example something analogous to a group association. That would allow libssl to distinguish between multiple TLS signature algorithm code points without requiring each code point to become a completely separate signature/keymgmt algorithm identity.

For GOST, that kind of model would be much easier for us to integrate with the current implementation structure.

We would very much appreciate any guidance on which direction would be more in line with the intended OpenSSL design here.

mattcaswell commented on ssl/t1_lib.c:
```diff
@@ -2394,15 +2394,20 @@ int ssl_setup_sigalgs(SSL_CTX *ctx)
+        size_t idx;
         TLS_SIGALG_INFO si = ctx->sigalg_list[i];
+
+        if (!ssl_cert_lookup_by_nid(OBJ_txt2nid(si.sigalg_name), &idx, ctx))
```
I'm suspicious of this. What if there is no nid for the sigalg_name?

VladGud replied:
As I understand it, I wouldn’t expect that to happen.

1) ctx->sigalg_list is built in add_provider_sigalgs().
2) There we call OBJ_create(sinf->sigalg_oid, sinf->sigalg_name, NULL).
3) We also check that OBJ_txt2nid(sinf->sigalg_name) is not NID_undef.

So when ssl_setup_sigalgs() reads si.sigalg_name, it should already have a valid NID.

But I may be missing something.

mattcaswell commented on ssl/t1_lib.c:
```diff
@@ -3550,6 +3554,19 @@ static int tls12_sigalg_allowed(const SSL_CONNECTION *s, int op,
         }
     }
 
+    /* TLS1.2 GOST sig algs could not be negotiated for the use in TLS1.3 */
```
```suggestion
    /* TLS1.2 GOST sig algs cannot be negotiated for use in TLS1.3 */
```

openssl-machine commented:
This PR has the label 'hold: cla required' and is stale: it has not been updated in 30 days. Note that this PR may be automatically closed in the future if no CLA is provided.  For CLA help see https://www.openssl.org/policies/cla.html

Commit:
Apply suggestions from code review

Add comment for early-data

Co-authored-by: Matt Caswell <matt@openssl.foundation>

reptiloidla commented:
```diff
diff --git a/ssl/t1_trce.c b/ssl/t1_trce.c
index e8075d0427..5f0a29b302 100644
--- a/ssl/t1_trce.c
+++ b/ssl/t1_trce.c
@@ -614,7 +614,15 @@ static const ssl_trace_tbl ssl_sigalg_tbl[] = {
      */
     { 0x0904, "mldsa44" },
     { 0x0905, "mldsa65" },
-    { 0x0906, "mldsa87" }
+    { 0x0906, "mldsa87" },
+    /* RFC9367 */
+    { 0x0709, "gostr34102012_256a" },
+    { 0x070a, "gostr34102012_256b" },
+    { 0x070b, "gostr34102012_256c" },
+    { 0x070c, "gostr34102012_256d" },
+    { 0x070d, "gostr34102012_512a" },
+    { 0x070e, "gostr34102012_512b" },
+    { 0x070f, "gostr34102012_512c" },
 };
 
 static const ssl_trace_tbl ssl_ctype_tbl[] = {
```

for correct name instead of UNKNOWN in signature_algorithms when using -trace

autoresolve from author

reptiloidla commented:
And for ciphersuites

```diff
diff --git a/ssl/t1_trce.c b/ssl/t1_trce.c
index e8075d0427..045b2904ad 100644
--- a/ssl/t1_trce.c
+++ b/ssl/t1_trce.c
@@ -453,6 +453,10 @@ static const ssl_trace_tbl ssl_ciphers_tbl[] = {
     { 0xC100, "GOST2012-KUZNYECHIK-KUZNYECHIKOMAC" },
     { 0xC101, "GOST2012-MAGMA-MAGMAOMAC" },
     { 0xC102, "GOST2012-GOST8912-IANA" },
+    { 0xC103, "TLS_GOSTR341112_256_WITH_KUZNYECHIK_MGM_L" },
+    { 0xC104, "TLS_GOSTR341112_256_WITH_MAGMA_MGM_L" },
+    { 0xC105, "TLS_GOSTR341112_256_WITH_KUZNYECHIK_MGM_S" },
+    { 0xC106, "TLS_GOSTR341112_256_WITH_MAGMA_MGM_S" },
 };
```

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 31 days ago

bob-beck commented:
So for doing this in TLS, the challenge today is that OpenSSL can support two PKEY based certs for TLS authentication. The reason for two is so "one can be EC and one can be RSA" - but this does not fit with GOST where it can sort of be many things.  

The reason GOST has been so painful to support is due to it not fitting well with OpenSSL's TLS story in the past and therefore requiring intrusive changes that both get in the way of other work, and which break frequently and make GOST a special flower - we don't want GOST to be a special flower, we just want it to be like any other algorithm. 

Some changes may be coming down the pipe soon that may make this a lot easier to deal with, and I think we should revisit this then, once we can perhaps have GOST dealt with "like other things" rather than requiring intrusive changes.

vdukhovni commented:
> So for doing this in TLS, the challenge today is that OpenSSL can support two PKEY based certs for TLS authentication. The reason for two is so "one can be EC and one can be RSA" - but this does not fit with GOST where it can sort of be many things.

Where "two" is to be read as "a few specific types", which now also include "ed25519", "ml-dsa" (each of the three sizes), and some others, but already with ECDSA there is only one slot for each of the various curves.
